### PR TITLE
Optimize boolean expressions & unwraps

### PR DESCRIPTION
**Pull request by eellison**
Simplify or eliminate boolean and/or expressions, optimize unwrapping a value that cannot be None, and optimize using `is` with a None and a non-None value

Since peephole optimize is now introducing constants, i added another constant propagation pass after running it.

Previously i had a PR that did this & optimized shape ops - i will add the shape optimizations in a separate PR.

### CODE REVIEW

File: torch/csrc/jit/passes/peephole.cpp [hunk 1]
```diff
@@ -165,6 +165,47 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
...
+            constant_as<bool>(n.thenOutputs().at(i)).value_or(false);
+        bool false_val =
+            constant_as<bool>(n.elseOutputs().at(i)).value_or(true);
+        // if an if node's output equals its condition replace output with
+        // condition
+        if (true_val && !false_val) {
+          n.outputs().at(i)->replaceAllUsesWith(n.cond());
+        }
+      }
+    } else if (
```
Comment (apaszke) [inline torch/csrc/jit/passes/peephole.cpp]: Can you please add a comment describing the optimization happening here?

File: torch/csrc/jit/passes/peephole.cpp [hunk 2]
```diff
@@ -165,6 +165,47 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
...
+        bool other_must_not_be_none = !other_input->mustBeNone() &&
+            !other_input->type()->cast<OptionalType>();
+        if (input_must_be_none && other_must_not_be_none) {
+          auto output = node->owningGraph()->insertConstant(
+              node->kind() == aten::__isnot__);
+          node->output()->replaceAllUsesWith(output);
+        }
+      }
+    } else if (
+        node->kind() == prim::unchecked_unwrap_optional ||
```
Comment (apaszke) [inline torch/csrc/jit/passes/peephole.cpp]: Same here

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -991,21 +992,28 @@ struct to_ir {
     const auto first_bool_info = findRefinements(first_expr);
     Value* first_value = emitCond(Expr(first_expr));
 
+    // if the condition is returned, then in an or expression the condition is
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: What's a "returned condition"? Can you please explain that better?

File: torch/csrc/jit/script/compiler.cpp [hunk 4]
```diff
@@ -2045,7 +2057,8 @@ struct to_ir {
         throw ErrorReport(loc) << "Only one argument to __new__ allowed";
       }
       return classNew->createObject(
-          apply.range(), method, Var(apply.inputs()[0]).name().name());;
+          apply.range(), method, Var(apply.inputs()[0]).name().name());
+      ;
```
Comment (suo) [inline torch/csrc/jit/script/compiler.cpp]: ?

File: torch/csrc/jit/script/schema_type_parser.cpp [hunk 5]
```diff
@@ -21,9 +21,17 @@ TypeAndAlias SchemaTypeParser::parseBaseType() {
...
       {"int", IntType::get()},
       {"bool", BoolType::get()},
+      {"None", NoneType::get()},
   };
-  auto tok = L.expect(TK_IDENT);
-  auto text = tok.text();
+  auto tok = L.cur();
+  std::string text;
+  if (L.nextIf(TK_NONE)) {
+    text = "None";
```
Comment (suo) [inline torch/csrc/jit/script/schema_type_parser.cpp]: Why is this necessary? Seems like the text of TK_NONE is "None" in lexer.h

File: torch/csrc/jit/passes/constant_propagation.cpp [hunk 6]
```diff
@@ -119,22 +120,40 @@ void inlineIf(Node* n, const AliasDb& aliasDb) {
...
-      false_block->eraseOutput(i);
-    } else {
-      i++; // increment bc we didn't remove current index
+      replaceAndRemoveIfOutput(n, i, true_block->outputs()[i]);
+      continue;
+    }
+
+    // true block output is constant and constant matches false block output
+    auto maybe_const = toIValue(t_out);
+    if (maybe_const && t_out->node() == f_out->node()) {
```
Comment (suo) [inline torch/csrc/jit/passes/constant_propagation.cpp]: having trouble understanding this one. If `t_out->node() == f_out->node()`, doesn't that mean that the node precedes the if statement (since it has to be accessible to both blocks)? If so, why did we ever emit if/else block outputs for it in the first place, and why do we have to insert a new constant, when presumably one already exists?

**Review by suo — APPROVED** [review]
looks generally fine, but a couple of cleanups/questions inline

**Comment by t-vi** [comment]
Just a quick shout that there *is* a bad interaction between this patch and my Optional specialization because the check used here does not consider is None to match the UndefinedTensor.
Of course you were right about cheking the function result in the unit test. :(

File: torch/csrc/jit/passes/peephole.cpp [hunk 7]
```diff
@@ -165,6 +165,51 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
...
+      // if we are comparing a None value with a value that can't be None
+      // replace the output with true if node is __isnot__ or false if node is
+      // __is__
+      AT_ASSERT(node->inputs().size() == 2);
+      for (size_t check_none_index : {0, 1}) {
+        bool input_must_be_none =
+            node->inputs().at(check_none_index)->mustBeNone();
+        Value* other_input = node->inputs().at(1 - check_none_index);
+        bool other_must_not_be_none = !other_input->mustBeNone() &&
+            !other_input->type()->cast<OptionalType>();
```
Comment (t-vi) [inline torch/csrc/jit/passes/peephole.cpp]: What is the role of `!other_intput->mustBeNone()` here?
I don't think that it is a good idea to use `! mustBeNone` as `cannotBeNone`.
Specifically, ! mustBeNone translates into producing node conditons of [`kind_ != prim::Constant || this->hasAttributes() || ! (output()->type()->cast<OptionalType>() || output()->type() == NoneType::get());`
](https://github.com/pytorch/pytorch/blob/1323c193edf97075bab63759cf0f353b7e7baa2d/torch/csrc/jit/ir.cpp#L773). In my opinion, the check should instead be for NoneType, OptionalType, and AutogradZeroTensorType (which might show up in backwards as None).
Maybe we should define a "canBeNone" helper function in jit_types.h.

As is, this will cause bugs in combination with specializing Optionals to None or AutogradZeroTensor.

File: torch/csrc/jit/passes/peephole.cpp [hunk 8]
```diff
@@ -165,6 +165,51 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
...
+          n.outputs().at(i)->replaceAllUsesWith(n.cond());
+        }
+      }
+    } else if (
+        node->kind() == aten::__is__ || node->kind() == aten::__isnot__) {
+      // if we are comparing a None value with a value that can't be None
+      // replace the output with true if node is __isnot__ or false if node is
+      // __is__
+      AT_ASSERT(node->inputs().size() == 2);
+      for (size_t check_none_index : {0, 1}) {
```
Comment (t-vi) [inline torch/csrc/jit/passes/peephole.cpp]: Why do you only check for must be none + must not be none and not for must be none and must be none?

File: torch/csrc/jit/passes/peephole.cpp [hunk 9]
```diff
@@ -165,6 +165,51 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
...
+          node->output()->replaceAllUsesWith(output);
+        }
+      }
+    } else if (
+        node->kind() == prim::unchecked_unwrap_optional ||
+        node->kind() == aten::_unwrap_optional) {
+      // we are unwrapping an input that can't be None, remove the unwrap
+      auto input = node->input();
+      if (input->type() != NoneType::get() &&
+          !input->type()->cast<OptionalType>()) {
```
Comment (t-vi) [inline torch/csrc/jit/passes/peephole.cpp]: Should this also cover AutogradZeroTensor?

File: torch/csrc/jit/passes/peephole.cpp [hunk 10]
```diff
@@ -165,6 +165,51 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
...
+      }
+    } else if (
+        node->kind() == aten::__is__ || node->kind() == aten::__isnot__) {
+      // if we are comparing a None value with a value that can't be None
+      // replace the output with true if node is __isnot__ or false if node is
+      // __is__
+      AT_ASSERT(node->inputs().size() == 2);
+      for (size_t check_none_index : {0, 1}) {
+        bool input_must_be_none =
+            node->inputs().at(check_none_index)->mustBeNone();
```
Comment (t-vi) [inline torch/csrc/jit/passes/peephole.cpp]: I think allowing None to be propagate from other nodes would be better, so here one would check with `mustBeNone` to cover optional, but also for `NoneType` and `AutogradZeroTensorType` separately.

**Review by t-vi — DISMISSED** [review]
I think the check for must not be none needs to be refined. The other two comments are less urgent, though. I'd think checking for `None is None` and refining the checks would be cool.
I'm not 100% certain about AutodiffZeroTensor, so you could also leave that out of the picture for comparing positive to None, but I think asserting they aren't none is difficult (and currently bugs with specialization).

In the end I guess #18360 is too risky a change, so you #18411 reverts it, you can ignore the comments.

File: torch/csrc/jit/script/compiler.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: Formatting change that i didn't look closely in, will remove extra `;`

File: torch/csrc/jit/script/schema_type_parser.cpp (hunk 5 above)
Comment (eellison) [reply torch/csrc/jit/script/schema_type_parser.cpp]: Previously the lexer was expecting an ident but found TK_NONE, causing an error. Yes instead of `text = "None"` I could write  `L.cur().text()` if that's what you're asking

File: torch/csrc/jit/passes/constant_propagation.cpp (hunk 6 above)
Comment (eellison) [reply torch/csrc/jit/passes/constant_propagation.cpp]: Agreed this is not very readable. I'm using the overloaded `==` from node_hashing, which checks that the Attributes are equal and that the output has the same type. I'll rewrite with method call. It doesn't mean that the node precedes the if statement. A constant node with the same attribute in each block is an example of when you would have to insert a new constant node before the if node.

File: torch/csrc/jit/passes/peephole.cpp (hunk 8 above)
Comment (eellison) [reply torch/csrc/jit/passes/peephole.cpp]: The two must be none cases is already handled by constant propagation. I'll add a comment

**Comment by eellison** [comment]
> I think the check for must not be none needs to be refined. The other two comments are less urgent, though. I'd think checking for `None is None` and refining the checks would be cool.
> I'm not 100% certain about AutodiffZeroTensor, so you could also leave that out of the picture for comparing positive to None, but I think asserting they aren't none is difficult (and currently bugs with specialization).
> 
> In the end I guess #18360 is too risky a change, so you #18411 reverts it, you can ignore the comments.

Thanks for your comments. I'm going to add a mustNotBeNone api, and modify mustBeNone to account for AutogradZero than I think this is ready to go

File: torch/csrc/jit/passes/peephole.cpp (hunk 8 above)
Comment (t-vi) [reply torch/csrc/jit/passes/peephole.cpp]: Ah, yes, but that is only for constants. :slightly_smiling_face:  Should we decide to specialize graphs based on `None` in `Optional[Tensor]` arguments, this would not be covered by const prop. Until then, I can see there is no need.

File: torch/csrc/jit/passes/constant_propagation.cpp (hunk 6 above)
Comment (eellison) [reply torch/csrc/jit/passes/constant_propagation.cpp]: Good catch - I was trying to use the Equal Node `==` operator but dont think I actually was.

File: torch/csrc/jit/passes/peephole.cpp (hunk 9 above)
Comment (eellison) [reply torch/csrc/jit/passes/peephole.cpp]: I talked with @wanchaol on Friday about this and IIRC he seemed to think mustBeNone doesn't need to check for AutogradZero. @wanchaol  could you weigh in?

File: torch/csrc/jit/passes/peephole.cpp (hunk 8 above)
Comment (eellison) [reply torch/csrc/jit/passes/peephole.cpp]: Up to this point if a node mustBeNone it's a constant. I still don't super understand the AutoGradZero case. If we can treat it as a constant I think we should modify constant propagation so that is the general behavior instead of just here @wanchaol

**Comment by eellison** [comment]
@t-vi take another look? I think it should be good to go now

File: torch/csrc/jit/passes/peephole.cpp (hunk 9 above)
Comment (t-vi) [reply torch/csrc/jit/passes/peephole.cpp]: As far as I could tell, AutogradZero is undefined tensors fed into `Optional[Tensor]` inputs for backward graphs.
I thought I saw a AutogradZero being treated as None somewhere, but I don't find it at the moment.

**Comment by t-vi** [comment]
The bits I know about seem OK.

File: torch/csrc/jit/script/compiler.cpp [hunk 11]
```diff
@@ -1039,21 +1039,32 @@ struct to_ir {
...
     } else {
+      first_value_returned = graph->insertConstant(false, nullptr, loc);
       first_expr_refinements = &first_bool_info.false_refinements_;
       second_expr_refinements = &first_bool_info.true_refinements_;
     }
 
     auto get_first_expr = [&] {
       insertRefinements(*first_expr_refinements);
-      return first_value;
+      return first_value_returned;
```
Comment (wanchaol) [inline torch/csrc/jit/script/compiler.cpp]: are we missing a replaceAllUsesWith the `first_value_returned`? we don't need to insertRefinements in these two case right?

File: torch/csrc/jit/passes/peephole.cpp [hunk 12]
```diff
@@ -165,6 +165,49 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
...
+          continue;
+        }
+        bool true_val =
+            constant_as<bool>(n.thenOutputs().at(i)).value_or(false);
+        bool false_val =
+            constant_as<bool>(n.elseOutputs().at(i)).value_or(true);
+        // if an if node's output equals its condition replace output with
+        // condition
+        if (true_val && !false_val) {
+          n.outputs().at(i)->replaceAllUsesWith(n.cond());
```
Comment (wanchaol) [inline torch/csrc/jit/passes/peephole.cpp]: what's the difference between this peephole optimization and the following `emitShortCircuitIf`? I feel like they are handling the same logic, and we should properly merge them.

File: torch/csrc/jit/ir.cpp [hunk 13]
```diff
@@ -771,9 +775,10 @@ bool Node::matches(
 }
 
 bool Node::mustBeNone() const {
-  return kind_ == prim::Constant && !this->hasAttributes() &&
-      (output()->type()->cast<OptionalType>() ||
-       output()->type() == NoneType::get());
+  return kind_ == prim::AutogradZero ||
```
Comment (wanchaol) [inline torch/csrc/jit/ir.cpp]: IIRC the only place where `AutogradZero` and `AutogradAdd` is added is when we specialize the inputs, and we don't need to handle this case in `unwrapoptional` related stuff if we don't specialize `Optional[Tensor]` to `AutogradZero` type (as that one need more discussions). but it might be ok for us to add a pre-check here. cc @zdevito

**Review by wanchaol — APPROVED** [review]
looks good, have some nits.

File: torch/csrc/jit/script/compiler.cpp (hunk 11 above)
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: At this point the values haven't been used yet, so there is no need to replace uses with

File: torch/csrc/jit/passes/peephole.cpp (hunk 12 above)
Comment (eellison) [reply torch/csrc/jit/passes/peephole.cpp]: Emit short circuit if isn't an optimization, it's emitting `expr-a and expr_b` by guarding the evaluation of `expr_b` in an if statement

File: torch/csrc/jit/ir.cpp (hunk 13 above)
Comment (eellison) [reply torch/csrc/jit/ir.cpp]: i think it would be safer & more future-proof to add the check

File: torch/csrc/jit/passes/peephole.cpp [hunk 14]
```diff
@@ -165,6 +165,49 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
...
+        }
+      }
+    } else if (
+        node->kind() == aten::__is__ || node->kind() == aten::__isnot__) {
+      // if we are comparing a None value with a value that can't be None
+      // replace the output with true if node is __isnot__ or false if node is
+      // __is__
+      AT_ASSERT(node->inputs().size() == 2);
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/peephole.cpp]: Probably don't _need_ to but if this had 3 inputs it would silently fail which is not good
